### ISSUE TITLE
Weights for unseen features.

### ISSUE THREAD

**Issue by swapniel99** [JohnLangford/vowpal_wabbit#988]
Hi,

How to set a weight for unseen feature encountered in test mode assuming there is hashing space available?

I was exploring --inital_weight x option.
For a small training dataset, I am seeing the intial weight x in most of the features using --readable_model. Seemed like it worked.

But for a large training set, I don't see it anymore. I have verified that the hashing space is not getting exhausted, because the number of lines(with weights) in readable model(say model A made using --initial_weight x) is 2^b and without --initial_weight option(say model B) it is much less.

When I searched for index which are not present in B but present in A, I saw weight 0. The training dataset used was same.

Please help.

**Comment by swapniel99** [comment]
After taking a higher value for b it is confirmed, that vw is giving 0 weights to unseen features:
Command used to train:
`cat trainingdata.txt | vw -b 20 --initial_weight -0.4411 --ignore x --loss_function=logistic --link=logistic --l1 0.0 --l2 1e-07 -f wtmodel.bin --readable_model wtmodel.txt`
Few lines of output model:
```
Version 8.1.1
Min label:-50.000000
Max label:50.000000
bits:20
lda:0
0 ngram:
0 skip:
options: --link=logistic
Checksum: -714986018
:0
0:-0.000000
1:-0.000000
2:0.000329
3:-0.000000
4:-0.000002
5:-0.000000
6:-0.000009
7:-0.000000
8:0.002628
9:-0.056540
10:-0.000000
11:-0.000000
12:-0.000000
```

Command used:
`cat trainingdata.txt | vw -b 20 --ignore x --loss_function=logistic --link=logistic --l1 0.0 --l2 1e-07 -f model.bin --readable_model model.txt`

```
Version 8.1.1
Min label:-50.000000
Max label:50.000000
bits:20
lda:0
0 ngram:
0 skip:
options: --link=logistic
Checksum: -714986018
:0
2:0.000329
3:-0.000000
4:-0.000002
5:-0.000000
6:-0.000013
8:0.002631
9:-0.056363
13:-0.001910
18:-0.000027
19:0.000315
21:-0.004675
```

**Comment by swapniel99** [comment]
Now I know I am getting 0 weights because of L2 regularisation.

How can I set L2 regularizer to take it to -0.4411 instead of 0?

**Comment by JohnLangford** [comment]
Hi Swapnil,

I'm not quite following your question, but what happens when you set the 
l2 regularizer to 0?

-John

**Comment by swapniel99** [comment]
When l2 regularizer is present the weights are suppressed so they get close to zero even if some data is present in training set, as it minimizes l2 norm ||w||_2 = ||w-0||_2

But I want to minimize ||w-c||_2 as regularization term where c is a constant vector other than 0, i.e. a vector containing constant c entirely.

**Comment by JohnLangford** [comment]
In the label field you can set the 'base' of the prediction to whatever 
you want (default 0).  Hence, you could set it to c * sum_i x_i and the 
prediction will be sum_i w_i x_i + c * sum_i x_i and the regularization 
will be purely on the w_i.

-John

**Comment by swapniel99** [comment]
Hi,

I did not understand. I simply need regularization to center around a constant non-zero value.
Currently a feature which occurs just a few times **in beginning** of huge training set will have 0 weight at the end of training. Regularization tries to tell that feature is unimportant so it gives it 0 weight. But I want unimportant features to have -ve weight.

Thanks.

**Comment by arielf** [comment]
Have you tried this option (appears in `vw --help`)?
```
  -C [ --constant ] arg           Set initial value of constant
```

**Comment by swapniel99** [comment]
Constant is just the bias term. I didn't mean that

**Comment by JohnLangford** [comment]
What you are looking for cannot be done.  The closest approximation is 
what I described.

-John
